File: modules/exomePeak/bin/geneId2name.py
import pandas as pd
import os
import pickle
import logging
from typing import Dict
logger = logging.getLogger(__name__)
def get_file_signature(file_path: str) -> Dict:
    """返回 GTF 文件签名，用于判断缓存是否失效"""
    stat = os.stat(file_path)
    return {
        "path": os.path.abspath(file_path),
        "size": stat.st_size,
        "mtime": stat.st_mtime,
    }


def save_cache(cache_path: str, gene_map: Dict[str, str], signature: Dict):
    """保存缓存：包括基因映射和 GTF 签名"""
    with open(cache_path, "wb") as f:
        pickle.dump({"signature": signature, "gene_map": gene_map}, f)


def load_cache_if_valid(cache_path: str, gtf_signature: Dict):
    """如果缓存存在且与当前 GTF 一致，则加载，否则返回 None"""
    if not os.path.exists(cache_path):
        return None

    try:
        with open(cache_path, "rb") as f:
            cache = pickle.load(f)
    except:
        return None  # 缓存损坏

    cached_sig = cache.get("signature", {})
    if cached_sig == gtf_signature:
        logger.info(f"cache matched, loading directly: {cache_path}")
        return cache["gene_map"]

    logger.info("cache mismatch detected → ignoring cache and rebuilding")
    return None


def parse_gtf_gene_map(gtf_path: str) -> Dict:
    """从 GTF 文件解析 gene_id → gene_name"""
    gene_map = {}

    with open(gtf_path, 'r') as f:
        for line in f:
            if line.startswith('#'):
                continue
            fields = line.strip().split('\t')
            if len(fields) < 9:
                continue
            if fields[2] != "gene":
                continue

            info = fields[8]
            attrs = {}

            for kv in info.split(';'):
                kv = kv.strip()
                if not kv:
                    continue
                parts = kv.replace('"', '').split(' ')
                if len(parts) == 2:
                    key, val = parts
                    attrs[key] = val

            gid = attrs.get("gene_id")
            gname = attrs.get("gene_name")

            if gid and gname:
                gene_map[gid] = gname

    logger.info(f"GTF 解析完成，共 {len(gene_map)} 个基因")
    return gene_map


def load_gtf_gene_map(gtf_path: str, cache_path="gene_map.pkl") -> Dict[str, str]:
    """
    Function to load gene_id to gene_name mapping from GTF file, with caching mechanism.
     - gtf_path: path to the GTF file
     - cache_path: path to the cache file (default: gene_map.pkl in the same directory as GTF)
     Returns a dictionary mapping gene_id to gene_name.
     The function first checks if a valid cache exists (matching GTF signature). If so, it loads the gene map from cache. Otherwise, it parses the GTF file to build the gene map, saves it to cache, and returns it.
    """
    gtf_signature = get_file_signature(gtf_path)

    gene_map = load_cache_if_valid(cache_path, gtf_signature)

    if gene_map is not None:
        return gene_map

    logger.info("begin parsing GTF to build gene map...")
    gene_map = parse_gtf_gene_map(gtf_path)

    cache_path = os.path.join(os.path.dirname(gtf_path), "gene_map.pkl")
    save_cache(cache_path, gene_map, gtf_signature)
    logger.info(f"cache written: {cache_path}")

    return gene_map


def translate_gene_ids(df:pd.DataFrame, gene_map: Dict[str,str],col:str):
    """将 Gene.refGene 列从 gene_id 转成 gene_name"""

    def convert(gene_ids):
        if pd.isna(gene_ids):
            return gene_ids
        ids = gene_ids.split(',')
        names = [gene_map.get(gid, gid) for gid in ids]
        return ",".join(names)

    return df[col].apply(convert)



def convert_gene_id_to_name(
    infile:str,
    gtf_path:str,
    gene_id_col:str="name",
    ):
    if not os.path.exists(infile):
        logger.error(f"输入文件不存在：{infile}")
        return None
    if not os.path.exists(gtf_path):
        logger.error(f"GTF 文件不存在：{gtf_path}")
        return None
    gtf_cache_path = os.path.join(os.path.dirname(infile), "gene_map.pkl")
    gene_map = load_gtf_gene_map(gtf_path, gtf_cache_path)
    df = pd.read_csv(infile, sep="\t")
    df["gene_name"] = translate_gene_ids(df, gene_map, gene_id_col)
    return df

if __name__ == "__main__":
    import argparse
    parser = argparse.ArgumentParser(description="translate gene_id to gene_name in a TSV file based on GTF annotation")
    parser.add_argument("--infile", required=True, help="input TSV file path (must contain a column with gene IDs)")
    parser.add_argument("--gtf", required=True, help="GTF file path")
    parser.add_argument("--outfile", required=True, help="output file path (TSV format)")
    parser.add_argument("--gene_id_col", default="name", help="column name for gene IDs (default: 'name')")
    args = parser.parse_args()

    result_df = convert_gene_id_to_name(args.infile, args.gtf, args.gene_id_col)
    if result_df is not None:
        result_df.to_csv(args.outfile, sep="\t", index=False)
        logger.info(f"转换完成，结果已保存：{args.outfile}")

File: modules/exomePeak/bin/geneId2name.r
#!/usr/bin/env Rscript
options(width = 60)

get_file_signature <- function(file_path) {
  info <- file.info(file_path)
  list(
    path = normalizePath(file_path, winslash = "/", mustWork = TRUE),
    size = info$size,
    mtime = as.numeric(info$mtime)
  )
}

save_cache <- function(cache_path, gene_map, signature) {
  saveRDS(list(signature = signature, gene_map = gene_map), cache_path)
}

load_cache_if_valid <- function(cache_path, gtf_signature) {
  if (!file.exists(cache_path)) {
    return(NULL)
  }
  cache <- tryCatch(readRDS(cache_path), error = function(e) NULL)
  if (is.null(cache)) {
    return(NULL)
  }
  cached_sig <- cache$signature
  if (is.null(cached_sig)) {
    return(NULL)
  }
  if (identical(cached_sig, gtf_signature)) {
    message("cache matched, loading directly: ", cache_path)
    return(cache$gene_map)
  }
  message("cache mismatch detected -> ignoring cache and rebuilding")
  NULL
}

parse_gtf_gene_map <- function(gtf_path) {
  gene_env <- new.env(hash = TRUE, parent = emptyenv())

  extract_attr <- function(info, key) {
    pattern <- paste0(key, " \\\"([^\\\"]+)\\\"")
    match <- regexpr(pattern, info, perl = TRUE)
    if (match[1] == -1) {
      return(NULL)
    }
    hit <- regmatches(info, match)
    sub(pattern, "\\1", hit, perl = TRUE)
  }

  con <- file(gtf_path, open = "r")
  on.exit(close(con), add = TRUE)

  repeat {
    lines <- readLines(con, n = 20000, warn = FALSE)
    if (length(lines) == 0) {
      break
    }
    for (line in lines) {
      if (startsWith(line, "#")) {
        next
      }
      if (!grepl("\tgene\t", line, fixed = TRUE)) {
        next
      }
      fields <- strsplit(line, "\t", fixed = TRUE)[[1]]
      if (length(fields) < 9) {
        next
      }
      info <- fields[9]
      gid <- extract_attr(info, "gene_id")
      gname <- extract_attr(info, "gene_name")
      if (!is.null(gid) && !is.null(gname)) {
        gene_env[[gid]] <- gname
      }
    }
  }

  gene_map <- as.list(gene_env)
  message("GTF parse done, genes: ", length(gene_map))
  gene_map
}

load_gtf_gene_map <- function(gtf_path, cache_path = "gene_map.rds") {
  gtf_signature <- get_file_signature(gtf_path)
  gene_map <- load_cache_if_valid(cache_path, gtf_signature)
  if (!is.null(gene_map)) {
    return(gene_map)
  }
  message("begin parsing GTF to build gene map...")
  gene_map <- parse_gtf_gene_map(gtf_path)
  save_cache(cache_path, gene_map, gtf_signature)
  message("cache written: ", cache_path)
  gene_map
}

translate_gene_ids <- function(value, gene_map) {
  if (is.na(value) || is.null(value)) {
    return(value)
  }
  value <- trimws(as.character(value))
  if (value == "") {
    return(value)
  }
  ids <- strsplit(value, ",", fixed = TRUE)[[1]]
  names <- vapply(ids, function(gid) {
    if (!is.null(gene_map[[gid]])) gene_map[[gid]] else gid
  }, character(1))
  paste(names, collapse = ",")
}

convert_gene_id_to_name <- function(infile, gtf_path, gene_id_col = "name") {
  if (!file.exists(infile)) {
    stop("input file not found: ", infile)
  }
  if (!file.exists(gtf_path)) {
    stop("GTF not found: ", gtf_path)
  }
  cache_path <- file.path(dirname(infile), "gene_map.rds")
  gene_map <- load_gtf_gene_map(gtf_path, cache_path)
  df <- read.table(infile,
                   sep = "\t",
                   header = TRUE,
                   quote = "",
                   comment.char = "",
                   stringsAsFactors = FALSE,
                   check.names = FALSE)
  if (!(gene_id_col %in% colnames(df))) {
    stop("input missing column: ", gene_id_col)
  }
  df$gene_name <- vapply(df[[gene_id_col]], translate_gene_ids, gene_map = gene_map, character(1))
  df
}

parse_args <- function() {
  args <- commandArgs(trailingOnly = TRUE)
  res <- list(infile = NULL, gtf = NULL, outfile = NULL, gene_id_col = "name")
  i <- 1
  while (i <= length(args)) {
    key <- args[[i]]
    if (key == "-i" || key == "--infile") {
      i <- i + 1
      res$infile <- args[[i]]
    } else if (key == "--gtf") {
      i <- i + 1
      res$gtf <- args[[i]]
    } else if (key == "-o" || key == "--outfile") {
      i <- i + 1
      res$outfile <- args[[i]]
    } else if (key == "--gene_id_col") {
      i <- i + 1
      res$gene_id_col <- args[[i]]
    }
    i <- i + 1
  }
  res
}

main <- function() {
  args <- parse_args()
  if (is.null(args$infile) || is.null(args$gtf) || is.null(args$outfile)) {
    stop("usage: geneId2name.R -i <infile> --gtf <gtf> -o <outfile> [--gene_id_col <col>]")
  }
  df <- convert_gene_id_to_name(args$infile, args$gtf, args$gene_id_col)
  write.table(df, file = args$outfile, sep = "\t", quote = FALSE, row.names = FALSE)
  message("done: ", args$outfile)
}

if (sys.nframe() == 0) {
  main()
}
